File: content/snippets/cpp/interview_lcci/05/02_binary_number_to_string.cpp
#include <iostream>
#include <string>

// Convert a decimal fraction (0 < n < 1) to its binary string representation.
// Return "ERROR" if the representation exceeds 32 characters.
std::string binaryToString(double n) {
    std::string result = "0.";
    while (n > 0) {
        if (result.size() > 32) return "ERROR";
        n *= 2;
        if (n >= 1) { result += '1'; n -= 1; }
        else result += '0';
    }
    return result;
}

int main() {
    std::cout << binaryToString(0.625) << '\n';
    std::cout << binaryToString(0.1) << '\n';
}

File: content/snippets/python/interview_lcci/05/02_binary_number_to_string.py
def binary_to_string(n: float) -> str:
    result = ["0."]
    while n > 0:
        if len(result) > 32:
            return "ERROR"
        n *= 2
        if n >= 1:
            result.append("1")
            n -= 1
        else:
            result.append("0")
    return "".join(result)


def main() -> None:
    print(binary_to_string(0.625))
    print(binary_to_string(0.1))


if __name__ == "__main__":
    main()
